lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -577,6 +584,83 @@ CompRegCanonicalizer::matchAndRewrite(seq::CompRegOp op,
   return success();
 }
 
+LogicalResult
+MergeVectorizeOps::matchAndRewrite(VectorizeOp vecOp,
+                                   PatternRewriter &rewriter) const {
+  auto &currentBlock = vecOp.getBody().front();
+  IRMapping argMapping;
+  SmallVector<Value> newOperands;
+  SmallVector<VectorizeOp> vecOpsToRemove;
+  bool canBeMerged = false;
+  // Used to calculate the new positions of args after insertions and removals
+  unsigned paddedBy = 0;
+
+  for (unsigned argIdx = 0, numArgs = vecOp.getInputs().size();
+       argIdx < numArgs; ++argIdx) {
+    auto inputVec = vecOp.getInputs()[argIdx];
+    // Make sure that the input comes from a `VectorizeOp`
+    // Ensure that the input vector matches the output of the `otherVecOp`
+    // Make sure that the results of the otherVecOp have only one use
+    // Ignore ClockTypes for now
+    auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
+    if (!otherVecOp || inputVec != otherVecOp.getResults() ||
+        !llvm::all_of(otherVecOp.getResults(),
+                      [](auto result) { return result.hasOneUse(); }) ||
+        llvm::any_of(otherVecOp.getBody().front().getArguments(), [](auto arg) {
+          return mlir::isa<circt::seq::ClockType>(arg.getType());
+        })) {
+      newOperands.insert(newOperands.end(), inputVec.begin(), inputVec.end());
+      continue;
+    }
+
+    canBeMerged = true;
+    newOperands.insert(newOperands.end(), otherVecOp.getOperands().begin(),
+                       otherVecOp.getOperands().end());
+
+    auto &otherBlock = otherVecOp.getBody().front();
+    for (auto otherArg : otherBlock.getArguments()) {
+      auto newArg = currentBlock.insertArgument(
+          argIdx + paddedBy, otherArg.getType(), otherArg.getLoc());
+      argMapping.map(otherArg, newArg);
+      ++paddedBy;
+    }
+
+    rewriter.setInsertionPointToStart(&currentBlock);
+    for (auto &op : otherBlock.without_terminator())
+      rewriter.clone(op, argMapping);
```

Comment:
Using `rewriter.inlineBlockBefore` would avoid cloning operations since the original ops are removed anyways.

Reply:
The body blocks don't have any successors since they are the only blocks in the region.
You'd need to modify the block arguments of the `otherVecOp` instead of the current block and pass the adequate values as the last argument of `inlineBlockBefore`. I don't think the code will get more complex by doing so, but it requires a few changes to your code. I'm also fine with leaving it as is for now. We can come back and get rid of the unnecessary cloning if we observe performance issues.